Paper: LabbookDB

TheChymera opened the pull request:
Authors: Horea-Ioan Ioanas, Markus Rudin

Abstract:
LabbookDB is a database application framework containing a relational database structure for life science research - along with a number of functions to conveniently add and retrieve information, and generate summaries. The core concept of LabbookDB is that most of the information commonly tracked in a lab book can be more efficiently and more reliably stored in a relational database. We overcome the portability limitations of designed-for-analysis spreadsheets and databases by building the database schema around atomized physical interactions of objects in the laboratory (and providing ready-for-plotting and/or ready-for-analysis dataframes as a compatibility layer). LabbookDB provides a wet work metadata storage model excellently suited for explorative ex-post reporting and analysis, as well as a potential infrastructure for automated wet work tracking.

Keywords:
laboratory notebook, labbook, wet work, record keeping, internet of things, reports, life science, biology, neuroscience, behaviour, relational database, normalization, SQL

stefanv commented:
Hi, @TheChymera, I'll be reviewing your submission; I'm looking forward to working with you on getting it included in the proceedings.

TheChymera commented:
Many thanks!

stefanv commented:
Latest PDF at: http://zibi.bids.berkeley.edu:7001/

stefanv commented:
I've reached out to some colleagues to see if we can also get a wet work domain expert to review.  I will take a look at the paper from a "scientific Python" perspective.

stefanv commented:
Thank you for your submission.  The paper is well written and clear.

I've left some minor editorial comments at:

[7.pdf](https://github.com/scipy-conference/scipy_proceedings/files/1086192/7.pdf)

In general, I think it would strengthen the paper if you could provide a stronger motivation for LabbookDB compared against a NoSQL dbs like Mongo.  Focus on where you add value above such general JSON data stores.  Find a killer feature that users *will not otherwise have*, and give an example.  Tell them how your package will help to improve their science.

Furthermore, please provide links to the software and its documentation, and mention whether you have quality control mechanisms (unit testing, etc.) in place (this will factor into scientists' decisions into whether the package can be trusted and adopted), and the development model (single person, lab, for how long will this be supported?).

If you have reached out to any other groups such as the MouseDB group, please include interesting bits from those conversations (e.g., does this package fill a niche, can the two teams' work complement on another, etc.)

Thank you, and I look forward to reading the next iteration!

manojpandey commented:
Hey @katyhuff  - I'll also be reviewing this submission !

flyingrobin commented:
Hi, I'll also be reviewing this submission.

flyingrobin commented:
Overall, the manuscript is very well written. I used to work in a biology lab and also worked with mice, so I can comment on the practical side of the software. 

Managing the mouse colony is a huge amount of work.  The proposed database system looks very interesting and promising to make the data recording and management much more efficient. However, I may want to add one point. I didn't see how users can easily add free text as an entry to the table, which is actually a very useful feature for the users. For example, in mouse colony management, researchers often need to write down some detailed observations, or some troubleshooting logs, so the protocol is definitely not enough for the framework of recording research activity. One analogy is the patient electronic record (ER), which is a complex database system in which the doctors have lots of freedom to log diagnosis details as free text. I would say the lab notebook is very much alike as the ER for patients. 

Although the manuscript focuses on the mouse colony management task, the authors did propose it as a general lab book tool. The mouse work is, after all, just one aspect of typical life science research routine. I'd like to see at least one more independent example of applications of LabbbokDB. A similar task could be the management of DNA sequences, such as primers, plasmid, or sequencing data.

deniederhut commented:
Hi @TheChymera ! I've been asked to provide a review for this submission.

Let me begin by saying that this is a really cool idea! I think many of us have struggled to convince our colleagues to move from flatfiles and paper into data management systems. My understanding is that labbookdb is intended as a usability layer over sql to make this kind of a transition painless. Neat! I do have some comments on the paper that I would like to see addressed. 

1. I would echo @stefanv 's request for a link to the code and documentation. 
2. Tracking a log of queries is an interesting way to version a database, but your RDBMS should have a way of dumping its data to plaintext, which can then be easily versioned by git. Maybe you have some functionality planned which is not supported by data dumps?
3. Maybe I missed this, but which RDBMS is being used by labbook? I'm guessing it is sqlite, based on the command syntax, but I don't recall seeing this written in the paper.
4. While CLIs are very flexible and easy to create, in general they tend to not promote discoverability for new users. Given that this software is intended to replace aspects of free form lab notebooks, I would like to ask if there are plans to create a graphical interface?
5. I am curious how flexible your table schemas are. As @flyingrobin pointed out, this tool is posited as a general purpose notebook replacement. How much work would it take to adapt labbookdb to track human behavioral studies? Polymer synthesis?

TheChymera commented:
Many thanks to all of you for your detailed feedback! The next revision addressing your comments and suggestions is in the making. Additionally, I would like to leave some comment-specific clarifications here:

@stefanv LabbookDB is not a DBMS (such as e.g. MySQL or MongoDB), but rather a schema with some application functionalities that integrate with the scientific Python stack. The underlying DBMS can be changed via SQLAlchemy, but in the current personal database incarnation of the project it is SQLite. The introduction explains the importance of relational modelling in laboratory metadata - and I can make the link between that and our DBMS choice more explicit.

@flyingrobin The database schema currently covers areas in which we perform numerous experiments: fMRI, neuroscience, optogenetics, behaviour as well as general mouse tracking. The scope has been expanding steadily as we incorporate metadata from collaborators, and we hope we can expand it even faster as we reach a broader audience of computationally-minded life scientists (including via SciPy 2017). There are already [some classes dealing with DNA manipulations](https://github.com/TheChymera/LabbookDB/blob/master/labbookdb/db/common_classes.py#L416), but they are not as fully developed (both schema and report-wise) as the behaviour and mouse monitoring work we show.

TheChymera commented:
Hello again everybody, I hope the replies offered here, as well as the additions to the paper and the package itself, adequately address the remaining open concerns!

stefanv commented on papers/horea-ioan_ioanas/paper.rst:
```diff
@@ -29,40 +34,47 @@ LabbookDB - A Wet-Work-Tracking Database Application Framework
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In order to cast complex laboratory metadata into tractable relationships with high enough entry numbers for statistical analysis, as well as in order to reduce data redundancy and the risk of introducing anomalies, we opt for a relational database management system, as interfaced with via SQLAlchemy.
+The scalability and flexibility advantages of flat data stores are not pertinent to the content at hand, as experimental metadata is comparatively smaller, more reliable, and more slowly obtained than data from other use cases which these systems were built to adress.
```
I'm not sure which "flexibility" advantages you refer to w.r.t. flat data stores (are you referring to flat file datastores?).  That seems to be irrelevant here.

My question of earlier, that was perhaps misunderstood, is why a more free-form schemaless storage mechanism, such as Mongo, would not work.  Here, I would presume the argument is because of query time on statistics, but then you mention here that you do *not* gather very large volumes of data at rapid pace.

TheChymera replied:
Not at all, given the data analysis applications at hand, query times are likely to remain entirely marginal concerns, even for somewhat more joins (though the growth of the schema will not necessarily increase joins for given use cases).

More free-form storage would not work (as efficiently and reliably), since it is precisely the relational nature of the DB which guarantees consistency over time (as mentioned, given the smaller volume of the data and its reliability, anomalies get to be more detrimental) and which makes sure that assignments such as that between a treatment group and a behavioural test evaluation can be performed without numerous entry duplication and consequently larger inputs.

Not least of all, a schema is needed to provide consistent query and output options. I imagine it could be done without a schema, but then you would be hard-coding an expected schema into the query, thus losing whatever input flexibility noSQL would have given you to begin with.

By flexibility I meant that noSQL databases like mongo could allow you to invent your attributes when you provide your input rather than when you design the storage.

stefanv commented on papers/horea-ioan_ioanas/paper.rst:
```diff
@@ -15,12 +19,13 @@ LabbookDB - A Wet-Work-Tracking Database Application Framework
-    We overcome the lack of flexibility and portability of designed-for-analysis spreadsheets and databases by building the database schema around atomized physical interactions of objects in the laboratory
-    (and providing ready-for-plotting and/or ready-for-analysis dataframes as a compatibility layer).
-    LabbookDB provides a wet work metadata storage model excellently suited for explorative ex-post reporting and analysis, as well as a potential infrastructure for automated wet work tracking.
+    The core concept of LabbookDB is that most information commonly tracked in a lab book can be more efficiently and more reliably stored in a relational database.
```
More efficiently and reliably than... ?  And what is meant by efficacy and reliability here?

TheChymera replied:
More efficiently than in a lab book, since that is where such information is commonly tracked.
By efficiency I mean each entry is put to more use, and by reliably I mean it guarantees better consistency. I guess some confusion might arise since these advantages are described in the article in the comparison between a relational database and a flat file store. And here in the comparison of a lab book and a relational DB. I will mention that in the abstract as well.

stefanv commented on papers/horea-ioan_ioanas/paper.rst:
```diff
@@ -29,40 +34,47 @@ LabbookDB - A Wet-Work-Tracking Database Application Framework
+
+In order to cast complex laboratory metadata into tractable relationships with high enough entry numbers for statistical analysis, as well as in order to reduce data redundancy and the risk of introducing anomalies, we opt for a relational database management system, as interfaced with via SQLAlchemy.
+The scalability and flexibility advantages of flat data stores are not pertinent to the content at hand, as experimental metadata is comparatively smaller, more reliable, and more slowly obtained than data from other use cases which these systems were built to adress.
+With an easily extendable but robust schema design we set out to reposit data in such a way that encapsulates contributors' wet work procedural knowledge in order to facilitate reproducibility and encourage standardization in reporting and comparability across experiments.
```
This sentence is densely packed with buzzwords.  Please either explain each sub-part or simplify.

TheChymera replied:
please see the latest revision.

stefanv commented on papers/horea-ioan_ioanas/paper.rst:
```diff
@@ -76,38 +88,62 @@ Record Keeping and Structure Migration
 This is an appropriate scope for most research groups.
+Aditionally, this design choice enables single researchers or clusters of researchers within a larger group to autonomously try out, test, contribute to, or adopt LabbookDB without significant overhead or the need for a larger institutional commitment.
+
+Quality Control
```
Thank you for adding these three sections---they satisfy my concerns.

TheChymera replied:
Thanks for helping me increase the paper's prospective impact ;)

stefanv commented:
@TheChymera Em-dashes are supposed to work like that :)

http://www.thepunctuationguide.com/em-dash.html

But, I'll defer to resident authority @mpacer

katyhuff commented:
Thank you for your reviews @stefanv @flyingrobin @deniederhut ! The authors seem to have made changes according to your comments. Do you feel this paper is now ready for inclusion in the SciPy proceedings?

stefanv commented:
@katyhuff I am satisfied with the paper, and look forward to seeing it appear in the proceedings.

mpacer commented:
I actually disagree with the space swallowing nature of that method of using an em-dash. But that's my personal opinion and is not a universal practice. But you could do it either way (including by including a unicode em-dash with spaces around it " — ".

But I'm also weird in that I think em- and en-dashes should be used for functionally different purposes in breaking up multi-clause sentences. Em-dashes in my mind are to be used when splitting a compound sentence into two related but individually complete chunks — to qualify for the use of an em-dash then, each clause should stand-alone as a complete sentence (similar to how one should use a semi-colon).  En-dashes in my mind are better for appositives – incomplete clauses in the middle of a sentence – where the en-dash indicates that the sentence's main content will resume after the end of the appositive.

flyingrobin commented:
The revision looks good to me! I appreciate the authors' work and hope it will benefit a large population of scientists!

mpacer commented:
2 out of 3 responses are sufficient in my book. I'm going to move from pending comment to ready. @deniederhut if you want to weigh in I'll wait until tomorrow morning before merging all the ready papers.

TheChymera commented:
@mpacer as a minor typesetting detail. What should I do about the em dash in the tytle? It seems the build system breaks as soon as I try to set it. For the moment it remains a hyphen.

mpacer commented:
@TheChymera Does that happen even if you use the Unicode character?

mpacer commented:
Also conventionally that'd be more likely to be a colon than an em-dash(with only a space following the colon). But I'm with you on the em-dash looking nicer in this case, likely with either a space preceding and postceding or not. So, I'd like to solve this if we can.

mpacer commented:
@deniederhut were your comments addressed? I see that the most recent comments to reviewers only addressed @stefanv & @flyingrobin.

mpacer commented:
Ok, merging so I can make the html version of the proceedings.
